### PR TITLE
fix(evaluation): SWE-bench evaluation script supports multiprocessing

### PR DESCRIPTION
**Pull request by csmith49**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR addresses two issues with evaluating SWE-bench results:

1. The `process_instance` function required more data than was possible to provide using the standard `run_evaluation` harness. This was previously resolved with the use of `global` variables, but each Python process has its own global namespace and so this would break when using multiple workers. Instead we pass the necessary data as an optional parameter and use partial evaluation to coerce the `process_instance_func` into the form that the evaluation harness expects.

2. The same call to `run_evaluation` is the only one in the codebase to not provide an `EvalMetadata` object. I'm in favor of making this required (we either need to generate it for the evaluation or it was already generated in a previous step) so I've rebuilt the metadata object and added a warning in `run_evaluation` when metadata is not provided.

---
**Link of any specific issues this addresses**

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:02b8b75-nikolaik   --name openhands-app-02b8b75   docker.all-hands.dev/all-hands-ai/openhands:02b8b75
```

### CODE REVIEW

File: evaluation/swe_bench/eval_infer.py [hunk 1]
```diff
@@ -94,13 +95,28 @@ def get_config(instance: pd.Series) -> AppConfig:
 
 def process_instance(
     instance: pd.Series,
-    metadata: EvalMetadata | None = None,
+    _metadata: EvalMetadata,
```
Comment (xingyaoww) [inline evaluation/swe_bench/eval_infer.py:98]: Is this change is necessary or not?

Also, we could probably feed this `metadata` to `EvalOutput` now?

Comment (csmith49) [reply evaluation/swe_bench/eval_infer.py:98]: > Is this change is necessary or not?

The parameter isn't used at the moment, so this isn't necessary for the script to work. A few reasons why I made the change:

1. We should always have a metadata object, so there's no reason to pass `None`.
2. `run_evaluation` expects a function of type `Callable[[pd.Series, EvalMetadata, bool], EvalData]`. If `run_evaluation` is following its type guarantees, this function will never be called with `None`. 

> Also, we could probably feed this `metadata` to `EvalOutput` now?

I hadn't seen that was a field on `EvalOutput`. I'll pass it through.

Comment (csmith49) [reply evaluation/swe_bench/eval_infer.py:98]: But I don't think point 2 is true. It'd be good to work towards type correctness when possible but happy to revert if you prefer.

Comment (xingyaoww) [reply evaluation/swe_bench/eval_infer.py]: Sounds good!

**Review by xingyaoww — APPROVED** [review]
Thanks and LGTM!

**Pull request merged**